File: apps/ios/AnpusFeather/Game/Music.swift
import AVFoundation
import CoreGraphics

/*
  Soundtrack — the port of music.pde.
  Copyright-free music "Anubis" by Lucha and R3VXS.
*/
final class Music {
    private var player: AVAudioPlayer?
    private var volume: Float = 1

    init(url: URL) {
        try? AVAudioSession.sharedInstance().setCategory(.ambient, options: [.mixWithOthers])
        try? AVAudioSession.sharedInstance().setActive(true)
        player = try? AVAudioPlayer(contentsOf: url)
        player?.numberOfLoops = -1
        player?.prepareToPlay()
    }

    func play() {
        volume = 1
        player?.volume = volume
        player?.play()
    }

    func fadeOut(_ dt: CGFloat) {
        guard let player, player.isPlaying, volume > 0 else { return }
        volume = max(0, volume - Float(0.05 * dt * FPS))
        player.volume = volume
    }

    func end() { player?.stop() }

    func toggleMute() {
        guard let player else { return }
        player.volume = player.volume > 0 ? 0 : volume
    }
}

File: apps/unity/Assets/AnpusFeather/Scripts/Music.cs
using System.Collections;
using UnityEngine;
using UnityEngine.Networking;

namespace AnpusFeather
{
    /*
      Soundtrack — port of music.pde.
      Copyright-free music "Anubis" by Lucha and R3VXS.
    */
    public sealed class Music
    {
        AudioSource _source;
        float _volume = 1f;
        bool _wantPlay;

        public bool IsReady => _source != null && _source.clip != null;

        public void Attach(AudioSource source) => _source = source;

        public IEnumerator Load(string path)
        {
            // UnityWebRequest needs a URI (file://) for local StreamingAssets paths.
            string url = path.Contains("://") ? path : new System.Uri(path).AbsoluteUri;
            using var req = UnityWebRequestMultimedia.GetAudioClip(url, AudioType.WAV);
            ((DownloadHandlerAudioClip)req.downloadHandler).streamAudio = true;
            yield return req.SendWebRequest();
            if (req.result != UnityWebRequest.Result.Success || _source == null) yield break;
            _source.clip = DownloadHandlerAudioClip.GetContent(req);
            _source.loop = true;
            _source.playOnAwake = false;
            if (_wantPlay) Play();
        }

        public void Play()
        {
            _wantPlay = true;
            if (_source == null || _source.clip == null) return;
            _volume = 1f;
            _source.volume = _volume;
            _source.Play();
        }

        public void FadeOut(float dt)
        {
            if (_source == null || !_source.isPlaying || _volume <= 0f) return;
            _volume = Mathf.Max(0f, _volume - 0.05f * dt * World.Fps);
            _source.volume = _volume;
        }

        public void End()
        {
            if (_source != null) _source.Stop();
        }

        public void ToggleMute()
        {
            if (_source == null) return;
            _source.volume = _source.volume > 0f ? 0f : _volume;
        }
    }
}
